src/lib/util/vec.ml

```diff
@@ -28,144 +28,144 @@
 (*                                                                        *)
 (**************************************************************************)
 
-type 'a t = { mutable dummy: 'a; mutable data : 'a array; mutable sz : int }
-
-let make capa d = {data = Array.make capa d; sz = 0; dummy = d}
+type 'a t = {
+  mutable data : 'a array;
+  mutable sz : int;
+  dummy: 'a;
+}
 
-let init capa f d =
-  {data = Array.init capa (fun i -> f i); sz = capa; dummy = d}
+let make n ~dummy = {data = Array.make n dummy; sz = 0; dummy}
+
+let[@inline] create ~dummy = {data = [||]; sz = 0; dummy}
+
+let[@inline] clear vec = vec.sz <- 0
+
+let[@inline] shrink vec i =
+  assert (i >= 0);
+  assert (i <= vec.sz);
+  for j = vec.sz - i to vec.sz - 1 do
+    Array.unsafe_set vec.data j vec.dummy
+  done;
+  vec.sz <- vec.sz - i
+
+let[@inline] pop vec =
+  assert (vec.sz > 0);
+  let x = Array.unsafe_get vec.data (vec.sz - 1) in
+  Array.unsafe_set vec.data (vec.sz - 1) vec.dummy;
+  vec.sz <- vec.sz - 1;
+  x
+
+let[@inline] last vec =
+  assert (vec.sz > 0);
+  Array.unsafe_get vec.data (vec.sz - 1)
+
+let[@inline] size vec = vec.sz
+
+let[@inline] is_empty vec = vec.sz = 0
+
+let[@inline] is_full vec = Array.length vec.data = vec.sz
+
+let[@inline] copy vec : _ t =
+  let data = Array.copy vec.data in
+  {vec with data}
+
+(* grow the array *)
+
+let[@inline never] grow_to vec cap : unit =
+  assert (Array.length vec.data < Sys.max_array_length);
+  let cap =
+    min Sys.max_array_length (max 4 cap)
+  in
+  let arr' = Array.make cap vec.dummy in
+  Array.blit vec.data 0 arr' 0 (Array.length vec.data);
+  vec.data <- arr';
+  assert (Array.length vec.data > vec.sz);
+  ()
+
+let[@inline never] grow_to_double_size vec : unit =
+  grow_to vec (2 * Array.length vec.data)
+
+let[@inline never] grow_to_by_double vec cap =
+  let cap = max 1 cap in
+  let c = ref (max 1 (Array.length vec.data)) in
+  while !c < cap do c := 2 * !c done;
+  grow_to vec !c
+
+let[@inline] push vec x : unit =
+  if is_full vec then grow_to_double_size vec;
+  Array.unsafe_set vec.data vec.sz x;
+  vec.sz <- vec.sz + 1
+
+let[@inline] get vec i =
+  assert (0 <= i && i < vec.sz);
+  let res = Array.unsafe_get vec.data i  in
+  if res == vec.dummy then raise Not_found;
+  res
+
+let[@inline] set vec i elt =
+  assert (not (elt == vec.dummy));
```

Comment:
`invalid_arg "Vec.set"` would be better

Reply:
The reason why we use assertions instead of exceptions is we can disable them for performance purposes.

Reply:
I don't think we actually do, but that makes sense.